pkg/store/limiter.go

```diff
@@ -0,0 +1,33 @@
+package store
+
+import (
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// Limiter is a simple mechanism for checking if something has passed a certain threshold.
+type Limiter struct {
+	limit uint64
+
+	// A ptr to a counter metric which we will increase if Check() fails.
+	failedCounter *prometheus.Counter
+}
+
+// NewLimiter returns a new limiter with a specified limit. 0 disables the limit.
+func NewLimiter(limit uint64, ctr *prometheus.Counter) *Limiter {
+	return &Limiter{limit: limit, failedCounter: ctr}
+}
+
+// Check checks if the passed number exceeds the limits or not.
+func (l *Limiter) Check(num uint64) error {
+	if l.limit == 0 {
+		return nil
+	}
+	if num > l.limit {
+		if l.failedCounter != nil {
+			(*l.failedCounter).Inc()
```

Comment:
Why casting?

Reply:
Will remove.

Reply:
Correction: failedCounter is a pointer so this needs to be derefenced before Inc() is called. We need a pointer here because in this way we decouple the counter from the metrics registry, and it makes it much easier to use in the future if the need will arise.

Reply:
got it know - we hold pointer to interface. It's interesting as this interface holds always *counter, so I think we are good without this extra ref pointer: `result := &counter{desc: desc, labelPairs: desc.constLabelPairs}`

Also can we get rid of nil checking everytime we do `Check`? I don't think it's necessary - I think we can ensure it's always set

Reply:
👍 yeah, we should just remove the nil checks since the caller ensures that. This happens also in some other places, will clean this up.